core/src/main/java/com/linecorp/armeria/internal/common/stream/EmptyFixedStreamMessage.java

```diff
@@ -16,19 +16,51 @@
 
 package com.linecorp.armeria.internal.common.stream;
 
+import static java.util.Objects.requireNonNull;
+
 import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+import org.reactivestreams.Subscriber;
 
 import com.google.common.collect.ImmutableList;
 
 import com.linecorp.armeria.common.annotation.Nullable;
 import com.linecorp.armeria.common.annotation.UnstableApi;
+import com.linecorp.armeria.common.stream.SubscriptionOption;
+import com.linecorp.armeria.common.util.UnmodifiableFuture;
+
+import io.netty.util.concurrent.EventExecutor;
 
 /**
  * A {@link FixedStreamMessage} that publishes no objects, just a close event.
  */
 @UnstableApi
 public class EmptyFixedStreamMessage<T> extends FixedStreamMessage<T> {
 
+    @Override
+    public boolean isComplete() {
+        return whenComplete().isDone();
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super T> subscriber, EventExecutor executor,
+                          SubscriptionOption... options) {
+        requireNonNull(subscriber, "subscriber");
+        requireNonNull(executor, "executor");
+        executor.execute(() -> {
+            subscriber.onSubscribe(NoopSubscription.get());
+            subscriber.onComplete();
+            whenComplete().complete(null);
+        });
+    }
+
+    @Override
+    public CompletableFuture<List<T>> collect(EventExecutor executor, SubscriptionOption... options) {
+        whenComplete().complete(null);
```

Comment:
Question) Is there no need to complete this future from the provided eventloop (`executor`)?
Ditto for `cancel`, `abort` as well.

Reply:
`executor` should be used to call the method on the subscriber not for completing a future.
If users want to execute the callback of a future with the specific executor, it's their obligation to use the executor
`future.handleAsync(..., executor)`

Reply:
I see. I understood that we don't have any contracts on which thread completes `whenComplete` 👍